modules/core/src/test/java/org/apache/ignite/internal/processors/cache/query/ScanIteratorTimeoutTest.java

```diff
@@ -0,0 +1,153 @@
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.apache.ignite.transactions.TransactionConcurrency;
+import org.apache.ignite.transactions.TransactionIsolation;
+import org.junit.Test;
+
+import javax.cache.Cache;
+import javax.cache.CacheException;
+import javax.cache.processor.EntryProcessor;
+import javax.cache.processor.EntryProcessorException;
+import javax.cache.processor.MutableEntry;
+import java.util.Iterator;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.apache.ignite.cache.CacheMode.PARTITIONED;
+
+/** Test */
+public class ScanIteratorTimeoutTest extends GridCommonAbstractTest {
+    /** {@inheritDoc} */
+    @Override protected void beforeTestsStarted() throws Exception {
+        super.beforeTestsStarted();
+
+        startGridsMultiThreaded(2, true);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String s) throws Exception {
+        return new IgniteConfiguration(super.getConfiguration(s)).
+                setCommunicationSpi(new BlockedTcpCommunicationSpi());
+    }
+
+    /** Test */
+    @Test
+    public void testTimeoutOnLoosingQueryResponse() throws Exception {
+        IgniteEx ignite0 = grid(0);
+        Ignite ignite1 = grid(1);
+
+        BlockedTcpCommunicationSpi commSpi0 = (BlockedTcpCommunicationSpi)ignite0.configuration().getCommunicationSpi();
+        commSpi0.blockFirstQueryResponse();
+
+        BlockedTcpCommunicationSpi commSpi1 = (BlockedTcpCommunicationSpi)ignite1.configuration().getCommunicationSpi();
+        commSpi1.blockFirstQueryResponse();
+
+        CacheConfiguration<Integer, String> cc = new CacheConfiguration<Integer, String>().
+                setName("cache").
+                setAtomicityMode(CacheAtomicityMode.TRANSACTIONAL).
+                setCacheMode(PARTITIONED);
+
+        IgniteCache<Integer, String> cache = ignite1.createCache(cc);
+
+        //writer
+        for (int i = 0; i < 1_000; i++)
+            cache.put(i, i + "_value");
+
+        IgniteInternalCache<Integer, String> cache1 = ignite0.cachex("cache");
+
+        try {
+            cache1.context().gate().enter();
+
+            try (GridNearTxLocal tx1 = cache1.txStartEx(TransactionConcurrency.PESSIMISTIC, TransactionIsolation.REPEATABLE_READ)) {
+                String val = cache1.get(34);
+                cache1.put(34, "NEW VAL");
+
+                Iterator<Cache.Entry<Integer, String>> itr = cache1.scanIterator(false, null, 2000);
+
+                while (itr.hasNext()) {
+                    Cache.Entry<Integer, String> entry = itr.next();
+
+                    cache1.invoke(entry.getKey(), new MyEntryProcessor());
+                }
+
+                tx1.commit();
+            } finally {
+                cache1.context().gate().leave();
+            }
+        }
+        catch (Throwable e) {
+            assertEquals(CacheException.class, e.getClass());
```

Comment:
Let's rewrite this with using GridTestUtils.assertThrows() and check exact exception type rather than just general CacheException

Reply:
ok